Add (failing) test for not hashing when we shouldn't

duog opened the pull request:
Hi,

I believe there is a bug in the logic for deciding whether to hash, and I believe these changes demonstrate it.

I've added a `shakeTraceHashFile :: FilePath -> FileHash -> IO ()` to `ShakeOptions` and I call this from inside the `unsafeInterleaveIO` that hashes the file. The digest test maintains an `IORef (HashMap FilePath FileHash)` and asserts it empty after build which should not hash anything(according to the comment).

This is a rough patch, but lmk what you think.

ndmitchell commented on src/Test/Digest.hs:
```diff
@@ -32,20 +37,25 @@ main = testBuild test $ do
+    clearTrackedHashes
     build ["--sleep","--digest-and"]
     assertOut "X"
+    assertEmptyHashes
```
So this is the failing step? The comment says there should be no hash, but you are saying this does perform a hash?

duog replied:
Yes that's right, the test fails here for me.

ndmitchell commented:
Thanks for spotting this! I think 458ecd5101cb4fbe3d3d1506a748283a6616adea fixes this. The issue was that anything in the RunResult is fully forced before being stored in the Shake database (which makes sense), but it noticed that the old and new were 100% equal, that nothing had changed, but then passed over the new value (which had the not-yet-forced thunks for hashes) rather than the old. I've fixed that and added a comment.

I'm not a massive fan of threading through a hook for hashing files, as that's pretty disruptive for a test. But at the same time, untested behaviour can easily regress...

duog commented:
> Thanks for spotting this! I think [458ecd5](https://github.com/ndmitchell/shake/commit/458ecd5101cb4fbe3d3d1506a748283a6616adea) fixes this. The issue was that anything in the RunResult is fully forced before being stored in the Shake database (which makes sense), but it noticed that the old and new were 100% equal, that nothing had changed, but then passed over the new value (which had the not-yet-forced thunks for hashes) rather than the old. I've fixed that and added a comment.
> 
> I'm not a massive fan of threading through a hook for hashing files, as that's pretty disruptive for a test. But at the same time, untested behaviour can easily regress...

I agree threading the hook through is gross. I'm sure your fix is good, will verify in a bit, but I'm dubious on the `EqualCost` stuff. For example, https://github.com/ndmitchell/shake/blob/458ecd5101cb4fbe3d3d1506a748283a6616adea/src/Development/Shake/Internal/Rules/File.hs#L162-L163 surely these lines should return `EqualExpensive`.

ndmitchell commented:
EqualExpensive means that the values are equal, but that it was more expensive than desired to compute that, so you should store a new entry in the database to make future equality cheaper. In both those cases it was expensive to compute, but not any more expensive than we would have expected, and storing a new entry in the database won't make future comparisons cheaper - so they are as cheap as they can be, even if that is still quite expensive.

duog commented:
> EqualExpensive means that the values are equal, but that it was more expensive than desired to compute that, so you should store a new entry in the database to make future equality cheaper. In both those cases it was expensive to compute, but not any more expensive than we would have expected, and storing a new entry in the database won't make future comparisons cheaper - so they are as cheap as they can be, even if that is still quite expensive.

:+1: that makes a lot of sense, thanks!

duog commented:
Closing, you fixed the bug (thanks!) and this test is too gross to be merged.

ndmitchell commented:
Given you found this in reality, would a release including this change be useful?

duog commented:
I found it hacking on hadrian. Yes, a release would be wonderful :) Thank
you

On Mon, Sep 6, 2021 at 9:28 PM Neil Mitchell ***@***.***>
wrote:

> Given you found this in reality, would a release including this change be
> useful?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ndmitchell/shake/pull/810#issuecomment-913842934>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2IKHLEFETPMWXOTP7AENDUAUQAFANCNFSM5DNJEMXQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

ndmitchell commented:
Released as 0.19.6

duog commented:
Hi, Just following up with some numbers. 

For me, a no-op build of `stage1:exe:ghc-bin` in hadrian speeds up ~ 50%, "+RTS -s" output attached.

I had a thought, perhaps rather than threading a hook through `ShakeOptions` we could report which files were actually hashed (by acting inside the unsafeInterleaveIO) in the profiling report?

[new.stat.txt](https://github.com/ndmitchell/shake/files/7128246/new.stat.txt)
[old.stat.txt](https://github.com/ndmitchell/shake/files/7128247/old.stat.txt)

ndmitchell commented:
Thanks for the numbers - that's great to see!

Recording which files were hashed in the profile report still requires threading that information somewhere, which is the ugly bit :(
